[mission001] todolist crud

StellaKim1230 opened the pull request:
#1 Mission001 

- [x] todo list에 todoItem을 키보드로 입력하여 추가
- [x] todo list의 체크박스를 클릭하여 complete 상태로 변경. (li tag 에 completed class 추가)
- [x] todo list의 x버튼을 이용해서 해당 엘리먼트를 삭제
- [x] todo list를 더블클릭했을 때 input 모드로 변경. (li tag 에 editing class 추가) 단 이때 수정을 완료하지 않은 상태에서 esc키를 누르면 수정되지 않은 채로 다시 view 모드로 복귀
- [x] todo list의 item갯수를 count한 갯수를 리스트의 하단에 보여주기
- [x] todo list의 체크박스를 클릭하여 complete 상태로 변경. (li tag 에 completed class 추가)

### 느낀점
해결하지 못한 이슈는 조금 더 고민해보고 작업해보겠습니다.
저는 class 를 사용하여 작업했는데, function을 사용했을때의 차이점과 각각에 대해서 더 알아야 겠다고 느꼈습니다.

StellaKim1230 committed:
할일 추가. (TODO: 추가시 input 초기화)
할일체크박스 선택. (TODO: li 처음 요소만 반응함)
할일 삭제 (TODO: li 처음 요소만 반응함)

StellaKim1230 committed:
WIP

StellaKim1230 committed:
WIP

StellaKim1230 committed:
WIP

StellaKim1230 committed:
WIP

StellaKim1230 committed:
WIP

Commit:
todoList 키보드로 입력하여 추가.
x 버튼 클릭시 할일 삭제.
체크박스 클릭시 클래스 completed 추가 구현.

Commit:
todo list의 count한 갯수를 리스트의 하단에 보여주기

StellaKim1230 committed:
Merge branch 'mission001/kimjieun' of https://github.com/EastjunDev/frontend into mission001/kimjieun

StellaKim1230 committed:
checkbox 유지.

StellaKim1230 committed:
할일 더블클릭시 input 으로 변경.

Commit:
코드 정리.

eastjun-dev commented on mission001/kimjieun/js/TodoInput.js:
```diff
@@ -0,0 +1,19 @@
+
+  init() {
+    this.$selector.addEventListener('keyup', (e) => {
+      if (e.keyCode === 13) {
```
`event.keyCode`같은 경우 [MDN 문서](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode)를 보시면 deprecated된 것을 확인하실 수 있어요~! [KeyboardEvent](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode)를 참고해서 업데이트 하시면 좋을 것 같아요 :)

StellaKim1230 replied:
deprecated 된지는 몰랐네요 ㅠㅠ KeyboardEvent 로 수정하겠습다 ㅎ

eastjun-dev commented on mission001/kimjieun/js/TodoInput.js:
```diff
@@ -0,0 +1,19 @@
+  init() {
+    this.$selector.addEventListener('keyup', (e) => {
+      if (e.keyCode === 13) {
+        const addData = {
```
이 오브젝트는 todoItem을 나타내는 것이기 때문에 `addData`라는 변수명이 오해를 만들 수 있을 것 같아요 :) 밑에 내용에서 이미 메서드로 `onAddTodo(data)`로 나타내기 때문에 `todoData` or `todoItem`으로 표현해도 괜찮지 않을까요?

eastjun-dev commented on mission001/kimjieun/js/app.js:
```diff
@@ -0,0 +1,57 @@
+
+  toggleTodo(target, dataIndex) {
+    this.data[dataIndex].isStatus = 'completed'
+    target.parentNode.setAttribute('class', 'completed')
```
- ui에 나중에 tag가 추가되는 경우가 발생할 수도 있을 것 같은데 parentNode보다는 [closest](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest)를 이용해보면 어떨까요~?
- class를 추가하는 경우 명시적인 `element.classList.add`가 있기에 해당 메서드를 쓰는게 의미가 더 명확할 것 같아요 :)

StellaKim1230 replied:
몰랐던 부분입니다!! 변경해보겠습니다! 감사합니다.

eastjun-dev commented on mission001/kimjieun/js/TodoInput.js:
```diff
@@ -0,0 +1,19 @@
+      if (e.keyCode === 13) {
+        const addData = {
+          text: e.target.value,
+          isStatus: '',
```
상태값을 나타내는 변수명이 `isStatus`인것보다는 명확하게 어떤 상태를 나타내는지 보여주면 더 좋을 것 같아요~!. class명에서도 `completed`로 style을 추가해주기 때문에 `isCompleted`와 같은 변수명으로 표현하면 어떨까요~?

StellaKim1230 replied:
isCompleted가 더 나을거 같네요 ㅠ 그때는 왜 isStatus밖에 떠오르지 않았는지.. ㅎ 리뷰 감사합니다

eastjun-dev commented on mission001/kimjieun/js/TodoList.js:
```diff
@@ -0,0 +1,32 @@
+
+  init() {
+    this.$selector.addEventListener('click', (e) => {
+      if (e.target.className === 'destroy') this.onDeleteTodo(e.target.parentNode.dataset.idx)
```
string으로 '직접' 비교하는 경우가 있는데, 해당 부분들을 따로 상수로 관리해 주시면 유지보수와, 협업 측면에서 상태를 추가 및 수정할 때 더 용이할 것 같아요 :)

eastjun-dev commented on mission001/kimjieun/js/TodoList.js:
```diff
@@ -0,0 +1,32 @@
+    })
+  }
+
+  createTodoListHtmlString = ({ text, isStatus }, index) => {
```
template code의 경우에는 indent를 맞춰주는게 구조를 볼 때 더 좋을 것 같아요 :)
```javascript
  createTodoListHtmlString = ({ text, isStatus }, index) => {
    return `<li ${isStatus ? `class=${isStatus}` : ''}>
              <div data-idx=${index} class="view">
                <input class="toggle" type="checkbox" ${isStatus === 'completed' && 'checked'}>
                <label class="label">${text}</label>
                <button class="destroy"></button>
              </div>
              <input class="edit" value="${text}">
            </li>`
  }
```

StellaKim1230 replied:
훨씬 가독성 있고 좋네요! 감사합니다.

eastjun-dev commented on mission001/kimjieun/js/TodoList.js:
```diff
@@ -0,0 +1,32 @@
+  </li>`
+  }
+
+  render = function(data) {
```
`createTodoListHtmlString`의 경우에는 arrow function으로 메서드를 표현하셨는데, `render`의 경우 함수 표현식으로 선언하신 이유가 있을까요? 일관된 함수 표기를 하면 협업할 때 더 좋을 것 같아요 :)

eastjun-dev commented on mission001/kimjieun/js/app.js:
```diff
@@ -0,0 +1,57 @@
+    this.todoList = todoList
+    this.todoInput = todoInput
+    this.todoCount = todoCount
+    this.$todoInputEl = document.querySelector('#new-todo-title')
```
변수명 앞에 $표시로 이미 dom element인걸 나타내기 때문에 변수 명 끝에 El은 없어도 괜찮지 않을까요? :)

eastjun-dev commented on mission001/kimjieun/js/index.js:
```diff
@@ -0,0 +1,20 @@
+import TodoList from './TodoList.js'
+import TodoInput from './TodoInput.js'
+import TodoCount from './TodoCount.js'
+
```
공백 정리를 해주시면 더 깔끔한 index.js가 될 것 같아요 :)

eastjun-dev reviewed (changes requested):
bind로 this를 넘겨줘서 깔끔하게 구현하신 app.js가 인상적이에요 ㅎㅎ
부분적으로 같이 참고해주셨으면 하는 부분 피드백 남겼습니다. 

그리고 커밋을 보내주실 때 💬 **커밋 메세지**💬 를 구체적으로 남겨주시면 더 좋을 것 같아요.
저는 개인적으로 [angularJS](https://gist.github.com/stephenparish/9941e89d80e2bc58a153)의 커밋컨벤션을 참고하는데, 취향에 맞으시는 규칙을 정해서 사용하시면 협업과, 개발 일정 및 계획세울 때도 도움이 되실 것 같아요 👍 

궁금하신 사항은 언제든지 코멘트를 통해 물어봐주세요. 😄

ganeodolu commented on mission001/kimjieun/js/app.js:
```diff
@@ -0,0 +1,57 @@
+  toggleTodo(target, dataIndex) {
+    this.data[dataIndex].isStatus = 'completed'
+    target.parentNode.setAttribute('class', 'completed')
+  }
```
`this.data[dataIndex].isStatus = this.data[dataIndex].isStatus === "completed" ? "":'completed';`
 ` target.parentNode.setAttribute('class', 'completed')`
    `this.render(this.data)`
toggle 버튼을 눌렀을 때 isState가 completed으로 고정되어서 다시 uncompleted 상태로 돌아오지 못하는 것 같아요. 버튼을 누르면 isState가 변경되고, 다시 render하는 방식으로 바꿔보았어요.
원하시는 답인지는 모르겠지만, 수정시에 참고하세요.

StellaKim1230 replied:
이 부분을 제가 놓치고 있었네요 ㅠ 감사합니다!

ganeodolu replied:
도움 될 수 있는 부분이 있어서 다행이에요. 수정하신 부분중에 this.data[dataIndex].isCompleted 값이 completed인지 new인지 비교하기 위해서도 그렇고 다른 부분에서도 동일한 변수를 비교할때 if, if를 주로 사용하셨는데 제 생각으로는 첫번째 조건을 통과해도 두번째 조건을 다시 비교해야하는데 if, else if를 사용하지 않고, if, if를 쓰신 이유가 있는지 궁금합니다.

StellaKim1230 replied:
@ganeodolu 개인적인 취향이지만, 저는 if ~ else if 구조를 최대한 피하는 편입니다. 말씀하신것처럼 첫번째 조건을 통과해도 두번째 조건을 다시 비교하는 것을 막기 위해선 return 을 해야하는데 제가 그부분은 놓친것 같네요. ( 좋은 지적 감사합니다 ^^ ) 👍 

예를들어 else if문 안에 또 다른 if 문이 중첩 되는 순간 로직이 머릿속에서 꼬이기 시작하여 최대한 담백하게  로직짜는 것을 좋아하고 else if 문을 사용 안 할수 있으면 최대한 사용 안합니다.

단, 어쩔수 없이 사용해야하는 부분은 사용을 합니다만..ㅠ 답변 도움이 되셨으면 좋겠네요^^

ganeodolu replied:
조건이 잘 구분되도록 else if를 피하는 것이었군요. 답변 감사합니다. :hushed:

YongHoonJJo commented on mission001/kimjieun/js/index.js:
```diff
@@ -0,0 +1,20 @@
+    }),
+    data,
+  })
+})()
```
이부분을 `new App({...})` 으로 하지 않고, 즉시 실행함수를 통해서 하신 이유가 따로 있으신가요?

StellaKim1230 replied:
즉시 실행해야 new 를 통해 인스턴스가 만들어지는줄 알았는데 딱히 그러지도 않네요 ^^;; 오버엔지니어링이라 코드 삭제하겠습니다. ㅎㅎ

YongHoonJJo commented on mission001/kimjieun/js/app.js:
```diff
@@ -0,0 +1,57 @@
+    this.todoCount = todoCount
+    this.$todoInputEl = document.querySelector('#new-todo-title')
+
+    this.data = data
```
아래 setState() 메서드에서 `this.data = data` 작업을 하기 때문에, 이 부분은 생략해도 되지 않을까요?

StellaKim1230 replied:
그러네요,, 중복으로 this.data = data를 하고있었네요 ^^; 삭제했습니다!

YongHoonJJo commented on mission001/kimjieun/js/app.js:
```diff
@@ -0,0 +1,57 @@
+    this.data = data
+    this.setState(data)
+
+    this.render()
```
이것도 역시 setState() 에서 `this.render()` 를 호출하기 때문에, 생략해도 되지 않을까요?

StellaKim1230 replied:
생략했습니다 ^^

YongHoonJJo commented on mission001/kimjieun/js/TodoCount.js:
```diff
@@ -0,0 +1,13 @@
+    return `총 <strong>${data.length}</strong> 개`
+  }
+
+  creatTodoCount = function(data) {
```
사소한거지만, 이 함수랑 위의 함수 모두 접두사가 `creat` 인데, 네이밍 오타인가요? ㅎㅎ..

StellaKim1230 replied:
@YongHoonJJo 네이밍 오타 맞습니다 ㅠ

YongHoonJJo commented on mission001/kimjieun/js/TodoList.js:
```diff
@@ -0,0 +1,32 @@
+
+  createTodoListHtmlString = ({ text, isStatus }, index) => {
+    return `<li ${isStatus ? `class=${isStatus}` : ''}>
+    <div data-idx=${index} class="view">
```
중간에 있는 데이터의 삭제가 발생할 경우, 위의 div 태그가 이전에 할당 받은 index 를 부여받는다는 보장이 없을 것 같은데, 이로 인해 문제가 발생할 여지는 없을까요??

StellaKim1230 replied:
그렇네요, 중간에 있는 데이터의 삭제가 발생할 경우 index 는 새로 부여받게 되는걸 확인했습니다~ 이 부분은 수정 해야 할 것 같네요 ^^ 감사합니다.

StellaKim1230 replied:
이전에 할당받은 Index 가 아니고, 새로할당 받은 index 여도 할일 리스트가 재랜더링 된 후에 해당 index 를 찾아가 삭제와 토글링 되니까 상관없지 않을까요?? 흠 좀 더 테스트 해봐야 할 것 같네요^^

이런 고민사항 좋습니다! ㅎ 감사합니다.

YongHoonJJo reviewed (commented):
웬만한건 동준님께서 자세히 리뷰해주신 것 같습니다. 저는 사소하지만 눈에 띄는 것들에 대해 코멘트 남겨봤습니다. :)
고생하셨어요!!

Commit:
add todo status filter render.

Commit:
코드리뷰 반영
1. 기능에 맞지 않는 변수명 수정
2. parentNode -> closest 으로 변경
3. string으로 직접 비교하는 부분 상수로 관리
4. template code indent 맞추기
5. 함수표현식, arrow function 중구난방으로 표현된 함수 arrow function으로 표현
6. 필요없는 코드 제거
7. fix typo

TODO
1. event.keyCode -> KeyboardEvent 로 변경
2. toggle -> completed 에서 uncompleted 로 돌아오지 않는 버그.

Commit:
add editing in constants file.

Commit:
리뷰반영 : keyCode 제거 및 keydown 으로 변경.

eastjun-dev reviewed (approved):
꼼꼼하게 피드백에 소통해주시고 반영해 주셔서 감동입니다 😭 
반영해주신점 확인하였고, approve하겠습니다 :)

StellaKim1230 committed:
리뷰반영
1. toggle 버튼클릭시 isCompleted 에서 new로 수정 안되는 버그.

ganeodolu commented on mission001/kimjieun/js/TodoCount.js:
```diff
@@ -0,0 +1,13 @@
+  }
+
+  createHtmlString = (data) => {
+    return `총 <strong>${data.length}</strong> 개`
```
저도 아직 구현 하지 않았지만, 해야할 일, 완료한 일 필터 버튼을 눌렀을 때 총 개수도 같이 변경되면 좋을 것 같아요.

StellaKim1230 replied:
@ganeodolu 꼼꼼한 리뷰 감사합니다! 해당 리뷰 적용해서 푸시했습니다^^

StellaKim1230 committed:
리뷰 반영
전체할일, 해야할일, 완료된일 클릭시 count 변경하는 로직 추가.

YongHoonJJo reviewed (approved):
늦어서 죄송합니다. ㅜㅜ
Approve 진행하겠습니다. 고생하셨어요!!

amorfati0310 commented on mission001/kimjieun/js/TodoCheck.js:
```diff
@@ -0,0 +1,16 @@
+
+  init = () => {
+    this.$selector.addEventListener('click', (e) => {
+      if (e.target.className === constant.ALLSELECTED) this.onTodoCheck(constant.ALL)
```
e.target.className 으로 분기 되는 부분이 같아서 switchCase 나 객체 프로퍼티로 분기하는  방식으로  빼주셔도 좋을 것 같아요 :)

StellaKim1230 replied:
가독성도 좋고, 필요없는 if 문 안 거쳐도 되어서 좋네요 ㅎㅎ 리뷰 감사합니다!

amorfati0310 commented on mission001/kimjieun/js/TodoCount.js:
```diff
@@ -0,0 +1,19 @@
+  }
+
+  createTodoCount = (status, data) => {
+    if (status === constant.ALL || !status) return this.$selector.innerHTML = this.createHtmlString(data)
```
status 가 없을때 default Parameter 로 constant.ALL 를 주는 것도 괜찮을 것 같네요

StellaKim1230 replied:
제 코드를 다시 보니깐, status 가 없을경우가 없네요 ㅎㅎ `|| !status` 제거했습니다. 
리뷰 감사합니다.^^

amorfati0310 commented on mission001/kimjieun/js/TodoCount.js:
```diff
@@ -0,0 +1,19 @@
+
+  createTodoCount = (status, data) => {
+    if (status === constant.ALL || !status) return this.$selector.innerHTML = this.createHtmlString(data)
+    if (status) {
```
status에 따라서 data 만 분기처리하고 있고   `this.$selector.innerHTML = this.createHtmlString(data)`
이 부분은 같아서 data만 변하게 하고 return  값은 같이 처리하는 건 어떨까요?

StellaKim1230 replied:
훨씬 코드가 깔끔해지네요 ㅎㅎ 감사합니다.

amorfati0310 commented on mission001/kimjieun/js/TodoInput.js:
```diff
@@ -0,0 +1,21 @@
+import { constant } from '../utils/constants.js'
```
constnant에서 가지고 오는 naming도 constant인데 조금 구체적인 todoStatus같은 네이밍으로 변경해도 좋을 것 같습니다 :)

amorfati0310 commented on mission001/kimjieun/js/TodoList.js:
```diff
@@ -0,0 +1,34 @@
+  }
+
+  createTodoListHtmlString = ({ text, isCompleted }, index) => {
+    return `<li class=${isCompleted}>
```
template 파일들을 따로 관리하는 것도 좋을 것 같아요 :)

amorfati0310 commented on mission001/kimjieun/js/app.js:
```diff
@@ -0,0 +1,72 @@
+  toggleTodo = (target, dataIndex) => {
+    if (this.data[dataIndex].isCompleted === constant.COMPLETED) {
+      this.data[dataIndex].isCompleted = constant.COMPLETED
+      target.closest('.completed').setAttribute('class', constant.NEW)
```
target.closest('.completed').setAttribute('class', constant.NEW)
 혹시 이렇게 바꾸면 중첩 class가 있을 때는 하나로 바뀌게 되는 경우가 있지 않을까요?

StellaKim1230 replied:
dataIndex로 해당 line을 찾아주니까 중첩 class 있어도 index로 찾아서 해당 엘리먼트만 바뀝니다!

amorfati0310 commented on mission001/kimjieun/js/app.js:
```diff
@@ -0,0 +1,72 @@
+      target.closest('.completed').setAttribute('class', constant.NEW)
+    }
+
+    if (this.data[dataIndex].isCompleted === constant.NEW) {
```
New , Completed 분기문에 따라 하는 동작이 유사해보이는데 하나의 함수로 합치는 것도 좋을 것 같아요

StellaKim1230 replied:
중복되는 부분 함수로 만들어서 처리하니 깔끔해졌네요 ㅎㅎ

amorfati0310 commented on mission001/kimjieun/js/app.js:
```diff
@@ -0,0 +1,72 @@
+  onTodoCheck = (status) => {
+    this.createTodoCount(status)
+    if (status === constant.ALL) return this.render(this.data)
+    const filteredData = this.data.filter((d) => d.isCompleted === status)
```
filter 함수에서 ({isCompleted}) destructuring 활용 하시면 더 좋을 것 같아요 
+ d라는 변수명 보다 의미를 더 드러내면 좋을 것 같습니다 :)

amorfati0310 commented on mission001/kimjieun/utils/constants.js:
```diff
@@ -0,0 +1,12 @@
+export const constant = {
```
상수따로 관리 좋네요 :)

amorfati0310 reviewed (commented):
리뷰가 도움이 되면 좋겠네요 수고 많으셨습니다 :)

ganeodolu reviewed (approved):
제가 클래스 방식이 생소해서 구조가 다 이해가 되지는 않았어도 기능별로 잘 구분이 되어서 깔끔해 보였습니다. 코드 보면서 많이 배우겠습니다. ^^

StellaKim1230 committed:
코드리뷰 반영.

ganeodolu commented on mission001/kimjieun/js/app.js:
```diff
@@ -0,0 +1,75 @@
+    target.closest('.new').setAttribute('class', todoStatus.EDITING)
+  }
+
+  onTodoCheck = (status) => {
```
해야할 일, 완료한 일 필터를 눌렀을 때 필터클래스에 selected를 넣어서 빨간 테두리가 유지되어있으면 더 좋을 것 같아요.
